libraries/AP_Soaring/AP_Soaring.cpp

```diff
@@ -0,0 +1,371 @@
+    // Start glide. Will be updated on the next loop.
+    _throttle_suppressed = true;
+}
+
+void SoaringController::get_wind_corrected_drift(const Location *current_loc, const Vector3f *wind, float *wind_drift_x, float *wind_drift_y, float *dx, float *dy)
+{       
+    Vector2f diff = location_diff(_prev_update_location, *current_loc); // get distances from previous update
+    *dx = diff.x;
+    *dy = diff.y;
+
+    // Wind correction
+    *wind_drift_x = wind->x * (AP_HAL::micros64() - _prev_update_time) * 1e-6;
+    *wind_drift_y = wind->y * (AP_HAL::micros64() - _prev_update_time) * 1e-6;
+    *dx -= *wind_drift_x;
+    *dy -= *wind_drift_y;
+}
+
+void SoaringController::get_altitude_wrt_home(float *alt)
+{
+    _ahrs.get_relative_position_D_home(*alt);
+    *alt *= -1.0f;   
+}
+void SoaringController::update_thermalling()
+{
+    struct Location current_loc;
+    _ahrs.get_position(current_loc);
+
+    if (_new_data) {  
+        float dx = 0;
+        float dy = 0;
+        float dx_w = 0;
+        float dy_w = 0;
+        Vector3f wind = _ahrs.wind_estimate();
+        get_wind_corrected_drift(&current_loc, &wind, &dx_w, &dy_w, &dx, &dy);
+
+        #if (0)
+            // Print32_t filter info for debugging
+            int32_t i;
+            for (i = 0; i < 4; i++) {
+                GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "%e ", (double)_ekf.P[i][i]);
+            }
+            for (i = 0; i < 4; i++) {
+                GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "%e ", (double)_ekf.X[i]);
+            }
+        #endif
+        
+        // write log - save the data.
+        DataFlash_Class::instance()->Log_Write("SOAR", "QfffffffLLfff", "TimeUS,nettorate,dx,dy,x0,x1,x2,x3,lat,lng,alt,dx_w,dy_w",
+                                              AP_HAL::micros64(),
+                                              (double)_vario_reading,
+                                              (double)dx,
+                                              (double)dy,
+                                              (double)_ekf.X[0],
+                                              (double)_ekf.X[1],
+                                              (double)_ekf.X[2],
+                                              (double)_ekf.X[3],
+                                              (double)current_loc.lat,
+                                              (double)current_loc.lng,
+                                              (double)_alt,
+                                              (double)dx_w,
+                                              (double)dy_w);
+            
+        //log_data(); 
+        _ekf.update(_vario_reading,dx, dy);       // update the filter
+         
+        _prev_update_location = current_loc;      // save for next time
+        _prev_update_time = AP_HAL::micros64();
+        _new_data = false;
+    }
+}
+
+void SoaringController::update_cruising()
+{
+    if (is_active() && suppress_throttle()) {
```

Comment:
My preference would be to reserve update_cruising() for future tasks that need to run continuously while in FBWB or AUTO mode, for example calculation of optimal airspeed and flap angle. init_cruising() is only ever run after checking (is_active() && suppress_throttle())  so these lines could be moved into init_cruising().